### ISSUE TITLE
Allow defaults for command-line options to be specified in a workflow

### ISSUE THREAD

**Issue by sappjw** [snakemake/snakemake#195]
Similar to #84, I would like my `snakemake` invocation to be the same by default for anybody who picks up the project after me. However, if I have specified (for example) a `--wrapper-prefix` it's not clear how to save that information in a reproducible way.  I don't think this makes sense for every option.  It may also make sense for it to be able to be overridden on the command line.

I could put it somewhere obvious, like in the `README`, but I'd rather have this automated.  I could also hard-code the URI in the rule configuration, for this specific option.

**Comment by troycomi** [comment]
I think you are looking for something like [profiles](https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles).  I usually wrap the invocation in a bash script so the options used are documented, but the profile is certainly cleaner.

**Comment by sappjw** [comment]
That's true, a profile stored with the analysis would work for this purpose.  However, without doing something as you suggested (make some sort of wrapper to Snakemake) the project wouldn't be able to stand by itself.

So an alternative (perhaps easier) option would be to allow for a configuration of the project profile directory within the Snakefile.

**Comment by troycomi** [comment]
I'm not sure that will be possible with the current framework but could be wrong.  By the time snakemake is reading the snakefile there are some options from the profile which would be needed.

Say you have a `profile` directive, similar to `configfile`, in your Snakefile.  In your profile, you could then have an entry to set the `snakefile` to `myworkflow.snake` which would be difficult to handle!  Depending on where the directive is located, other options could alter execution, such as allow-ambiguous.

An alternative would be to have a default profile filename that snakemake looks for if none is specified (like how Snakefile is handled).  I could see such a change breaking some workflows if the name is `profile.yaml` or similar.  Maybe your enhancement request should be 'Add default profile'?

I understand the desire to encapsulate complexity from users to some extent, but I'm assuming they will already have to alter either the config yaml or a working directory argument to get your workflow to run on their system.  Adding the direction to run snakemake with a profile option or use the wrapper you provide doesn't seem that onerous.  I'm not sure what your desire is to 'stand by itself'.

**Comment by sappjw** [comment]
The use case I'm thinking of is if a team member who uses the same compute environment has to pick up where I left off, I want to make it as seamless and easy as I can.  Currently there's no need for somebody else to modify the config files of the project(s) I'm working on, but I could see that being a common scenario.  For me it's just `git clone` and `snakemake` to build all the targets (yes, the latter command could be a shell script wrapper).

I think what I'm really suggesting is that the `--wrapper-prefix` be able to be stored in a Snakemake config file (I didn't see any others that make sense), so I will update the title.  It would be similar to `--configfile`, I think.

The ability to specify different wrapper prefixes on the command line doesn't seem to fit with all the built-in reproducibility that Snakemake already has.  It's related to the workflow and not the workflow environment, as the other command-line options are.  If I could specify this in a config file, I wouldn't have to document it elsewhere (by making a wrapper around `snakemake` or writing instructions).

All that said, I've thought of another alternative that is seamless: add the wrapper repository as a git submodule in my project directory.  So feel free to close.

**Comment by m-jahn** [comment]
- for these "environment"-type options, [profiles](https://snakemake.readthedocs.io/en/latest/executing/cli.html#profiles) are the way to go
- the `config.yaml` file could also be an option but I'd think this is more meant to set the options for the workflow behavior itself, not for other things like compute resources, cores, special paths, env vars etc

Profiles are now a pretty accepted and widely used way to configure the (local) environment, so this issue can be closed.
